### PR TITLE
Upweight relevant core tags in bottom of post recent opportunities

### PR DESCRIPTION
**Pull request by Will-Howard**
See [designs](https://www.figma.com/file/IRUVQ5D2mKxoW0xZnhXIWv/2024-Q1?type=design&node-id=485-8062&mode=design&t=4xXmQjhfRKmlEvVV-0) here. You can look at [this post](http://localhost:3000/posts/BvXkG3PLfdmvoECFb/this-is-why-we-can-t-have-nice-laws) as an example where the label with "in Animal Welfare" shows up.

I wanted to make it so if there are no good opportunities in the relevant core tag it falls back to other opportunities, so I implemented it as a karma multiplier (by a factor of 30) for posts that have a core tag the same as the post you are on. If there are enough of these that the top three results all have the same core tag it will say "Recent opportunities in X", otherwise it will still say "Recent opportunities" (but it may be the case that one or two of the results have been bumped up by this karma modifier, but there wasn't a third with the right tag). cc @agnesstenlund , in the case where there is no X do you think it should say something like "Relevant opportunities" instead, as it is still upweighting based on the tags?

If the post you are on has _multiple_ core tags (which is reasonably common), then these multipliers will stack (so a post with two matching core tags will have a multiplier of 30*30 = 900, which in practice means a post with >0 karma will probably make it to the top of the list). If all three returned posts match two or more core tags it will just pick the lowest alphabetically for the X to show in "Recent opportunities in X".

Screenshot 2024-03-04 at 13 35 46

---

I'm not sure about this multiple tag behaviour, i.e. giving much higher weight to posts that share two or more core tags. About 60% of posts in the last year do have two or more core tags:

Screenshot 2024-03-04 at 13 58 26

I ran a query to get a rough idea of:
- How many posts would be able to fill the recommendations with posts sharing one core tag
- How many would be able to fill the recommendations with posts sharing two or more core tags

Of the 385 opportunities posts in the last year:
- 116 (42%) could fill the section with posts with two shared core tags
- 246 (90%) could fill the section with posts with one shared core tag

I'm not 100% sure this query is right, it is consistent with the number above though (which show that a surprising number of posts have multiple core tags)

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1206753250297296) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/lib/collections/posts/views.ts [hunk 1]
```diff
@@ -419,12 +419,18 @@ function filterModeToAdditiveKarmaModifier(mode: FilterMode): number {
 }
 
 function filterModeToMultiplicativeKarmaModifier(mode: FilterMode): number {
-  if (typeof mode === "number" && 0 < mode && mode < 1) {
+  // Example: "x10.0" is a multiplier of 10
+  const match = typeof mode === "string" && mode.match(/^x(\d+(?:\.\d+)?)$/);
```
Comment (Will-Howard) [inline packages/lesswrong/lib/collections/posts/views.ts:423]: cc @b0b3rt for awareness, in the long run I would actually like to get rid of the rule where between 0 and 1 means it's multiplicative

**Comment by agnesstenlund** [comment]
Thanks so much for this, and for the careful thought going into it 🙏 

I agree "Relevant opportunities" or perhaps even better "Related opportunities" would be good, although I worry about both of them being a bit too presumptuous (you could imagine the opportunity not actually feeling very related, and that throwing people off). In that case we could also stick with "Recent opportunities". Do you have a preference?

**Comment by agnesstenlund** [comment]
Nice let’s do that. When it’s only one tag we should probably keep the “Recent opportunities in x” though, since they are not actually particularly relevant WITHIN x, but just recent? On 6 Mar 2024, at 18:11, Will Howard ***@***.***> wrote:﻿
@agnesstenlund
I think "Related opportunities" is good, I think people will be used to recommendations sometimes being not great so won't mind that much if they're not that related in some cases. I'll change it to "Related opportunities" and (e.g.) "Related opportunities in Animal Welfare" for the case where there is only one tag

File: packages/lesswrong/components/hooks/useRecentOpportunities.ts [hunk 2]
```diff
@@ -15,33 +17,73 @@ const subscribedTags: string[] = [
...
   maxAgeInDays?: number,
-}): UseMultiResult<FragmentTypeName> => {
+  post?: PostsWithNavigation | PostsWithNavigationAndRevision,
+}): UseMultiResult<FragmentTypeName> & {coreTagLabel: string | null} => {
+  const coreTags =
+    post?.tags
+      ?.filter((tag) => tag.core && !requiredTags.includes(tag._id))
+      .sort((a, b) => (a.name < b.name ? -1 : 1)) ?? [];
+  const coreTagIds = coreTags?.map(tag => tag._id) ?? [];
+  const subscribedTags = difference(defaultSubscribedTags, coreTagIds);
```
Comment (s-cheng) [inline packages/lesswrong/components/hooks/useRecentOpportunities.ts:42]: I'm not sure I understand the purpose of this line? I think you could just use `defaultSubscribedTags` directly below?

**Review by s-cheng — APPROVED** [review]
LGTM thanks!

File: packages/lesswrong/components/hooks/useRecentOpportunities.ts (hunk 2 above)
Comment (s-cheng) [reply packages/lesswrong/components/hooks/useRecentOpportunities.ts:42]: Oh I see what you mean, sounds good 👍

**Pull request merged**